x-pack/plugins/apm/server/lib/alerts/chart_preview/get_transaction_duration.ts

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { MetricsAggregationResponsePart } from '../../../../../../typings/elasticsearch/aggregations';
+import {
+  PROCESSOR_EVENT,
+  SERVICE_NAME,
+  TRANSACTION_DURATION,
+  TRANSACTION_TYPE,
+} from '../../../../common/elasticsearch_fieldnames';
+import { ProcessorEvent } from '../../../../common/processor_event';
+import { rangeFilter } from '../../../../common/utils/range_filter';
+import { AlertParams } from '../../../routes/alerts/chart_preview';
+import { getEnvironmentUiFilterES } from '../../helpers/convert_ui_filters/get_environment_ui_filter_es';
+import { getBucketSize } from '../../helpers/get_bucket_size';
+import { Setup, SetupTimeRange } from '../../helpers/setup_request';
+
+export async function getTransactionDurationChartPreview({
+  alertParams,
+  setup,
+}: {
+  alertParams: AlertParams;
+  setup: Setup & SetupTimeRange;
+}) {
+  const { apmEventClient, start, end } = setup;
+  const {
+    aggregationType,
+    environment,
+    serviceName,
+    transactionType,
+  } = alertParams;
+
+  const query = {
+    bool: {
+      filter: [
+        { range: rangeFilter(start, end) },
+        { term: { [PROCESSOR_EVENT]: ProcessorEvent.transaction } },
+        ...(serviceName ? [{ term: { [SERVICE_NAME]: serviceName } }] : []),
+        ...(transactionType
+          ? [{ term: { [TRANSACTION_TYPE]: transactionType } }]
+          : []),
+        ...getEnvironmentUiFilterES(environment),
+      ],
+    },
+  };
+
+  const { intervalString } = getBucketSize({ start, end, numBuckets: 20 });
+
+  const aggs = {
+    timeseries: {
+      date_histogram: {
+        field: '@timestamp',
+        fixed_interval: intervalString,
+      },
+      aggs: {
+        agg:
+          aggregationType === 'avg'
+            ? { avg: { field: TRANSACTION_DURATION } }
+            : {
+                percentiles: {
+                  field: TRANSACTION_DURATION,
+                  percents: [aggregationType === '95th' ? 95 : 99],
+                },
+              },
+      },
+    },
+  };
+  const params = {
+    apm: { events: [ProcessorEvent.transaction] },
+    body: { size: 0, query, aggs },
+  };
+  const resp = await apmEventClient.search(params);
+
+  if (!resp.aggregations) {
+    return [];
+  }
+
+  return resp.aggregations.timeseries.buckets.map((bucket) => {
+    const percentilesKey = aggregationType === '95th' ? '95.0' : '99.0';
+    const x = bucket.key;
+    const y =
+      aggregationType === 'avg'
+        ? (bucket.agg as MetricsAggregationResponsePart).value
+        : (bucket.agg as { values: Record<string, number | null> }).values[
+            percentilesKey
```

Comment:
Are the explicit types necessary?

Reply:
Maybe? If you take the first one off you get

```
Property 'value' does not exist on type 'MetricsAggregationResponsePart | { values: Record<string, number | null>; }'.
  Property 'value' does not exist on type '{ values: Record<string, number | null>; }'.ts(2339)
```

and if you take the second off you get 


```
Property 'values' does not exist on type 'MetricsAggregationResponsePart | { values: Record<string, number | null>; }'.
  Property 'values' does not exist on type 'MetricsAggregationResponsePart'.
```

so if there's a better way to narrow it earlier in the query i'd like to do that but this worked and made sense to me given the errors when it's not there.